From: Vanadiae <vanadiae35@gmail.com>
Date: Fri, 29 May 2020 00:44:04 +0200
Subject: Use the application name in the print notification (#304)

When printing with an flatpak application, the notification
telling that it prints says "Document from %s" where %s is the
app desktop file id (i.e. will print "Document from
org.gnome.Evince" with evince).

This is not very clear what this name is in the notification,
particularly with a third-party app, for example
com.github.my_beautiful_github_name.app_name.

So this changes %s to the actual translated application name,
looking in the corresponding desktop file entry.

Origin: upstream, 1.7.2, commit:68c044c2c97261baa9b154ad29ab3ec826dd26b4
---
 src/print.c | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/src/print.c b/src/print.c
index 411de4c..ce4d435 100644
--- a/src/print.c
+++ b/src/print.c
@@ -31,6 +31,7 @@
 #include <gio/gunixfdlist.h>
 #include <gio/gunixinputstream.h>
 #include <gio/gunixoutputstream.h>
+#include <gio/gdesktopappinfo.h>
 
 #include "xdg-desktop-portal-dbus.h"
 
@@ -259,8 +260,27 @@ print_file (int fd,
   g_autoptr(GUnixInputStream) istream = NULL;
   g_autoptr(GUnixOutputStream) ostream = NULL;
   int fd2;
+  g_autofree char *app_desktop_fullname = NULL;
+  g_autoptr (GDesktopAppInfo) app_info = NULL;
+  // ensures the app_name won't be NULL even if the app_id.desktop file doesn't exist
+  g_autofree char *app_name = g_strdup (app_id);
 
-  title = g_strdup_printf ("Document from %s", app_id);
+  if (!g_str_has_suffix (app_id, ".desktop"))
+    {
+      app_desktop_fullname = g_strconcat (app_id, ".desktop", NULL);
+    }
+  else
+    {
+      app_desktop_fullname = g_strdup (app_id);
+    }
+
+  app_info = g_desktop_app_info_new (app_desktop_fullname);
+  if (app_info)
+    {
+      app_name = g_desktop_app_info_get_locale_string (app_info, "Name");
+    }
+
+  title = g_strdup_printf ("Document from %s", app_name);
 
   if (!preview)
     job = gtk_print_job_new (title, printer, settings, page_setup);
